File: four_rings_java/src/fourrings/RingNode.java
package fourrings;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class RingNode extends Thread {

    private final String ringId;
    private final RingManager manager;

    private RingNode next;

    private final BlockingQueue<Token> mailbox = new LinkedBlockingQueue<>();

    public RingNode(String ringId, RingManager manager) {
        this.ringId = ringId;
        this.manager = manager;
    }

    public void setNext(RingNode next) {
        this.next = next;
    }

    public void send(Token token) throws InterruptedException {
        mailbox.put(token);
    }

    private long transform(long v) {

        switch (ringId) {

            case "NEG":
                return Math64.negTransform(v);

            case "ZERO":
                return Math64.zeroTransform(v);

            case "POS_EVEN":
                return Math64.posEvenTransform(v);

            default:
                return Math64.posOddTransform(v);
        }
    }

    public void run() {

        try {

            while (true) {

                Token token = mailbox.take();

                token.currentVal = transform(token.currentVal);

                token.remainingHops--;

                if (token.remainingHops > 0) {

                    next.send(token);

                } else {

                    manager.tokenFinished(token);
                }
            }

        } catch (InterruptedException e) {
            // shutdown
        }
    }
}

File: four_rings/lib/four_rings/ring_node.ex
defmodule FourRings.RingNode do
  @moduledoc """
  A single node in a ring.
  """

  alias FourRings.Math64

  def start(ring_id, manager_pid) do
    spawn(fn -> wait_for_next(ring_id, manager_pid) end)
  end

  defp wait_for_next(ring_id, manager_pid) do
    receive do
      {:set_next, next_pid} ->
        loop(ring_id, manager_pid, next_pid)

      :stop ->
        :ok
    end
  end

  defp loop(ring_id, manager_pid, next_pid) do
    receive do
      {:token, token} ->
        new_val = transform(ring_id, token.current_val)

        updated_token = %{
          token
          | current_val: new_val,
            remaining_hops: token.remaining_hops - 1
        }

        if updated_token.remaining_hops > 0 do
          send(next_pid, {:token, updated_token})
        else
          send(manager_pid, {:token_complete, updated_token})
        end

        loop(ring_id, manager_pid, next_pid)

      :stop ->
        :ok
    end
  end

  defp transform(:neg, v), do: Math64.add(Math64.mul(v, 3), 1)
  defp transform(:zero, v), do: Math64.add(v, 7)
  defp transform(:pos_even, v), do: Math64.mul(v, 101)
  defp transform(:pos_odd, v), do: Math64.add(Math64.mul(v, 101), 1)
end
